osd: full-object read CRC mismatch due to 'truncate' modifying oi.size w/o clearing 'data_digest'

NitzanMordhai opened the pull request:
This is continue of PR:
https://github.com/ceph/ceph/pull/43902
the commits cherry-picked and i added some tests that recreate the issue without the fix

when write before truncate, need trim length, if truncate is to 0,
write is [0128k], write change to [00], do nothing, oi.size is 0, x1 = set_data_digest(crc32(-1)).
write is [128k128k], write change to [128k0], truncate oi.size to offset 128k, x2 = set_data_digest(crc32(x1)).
write is [256k128k], write change to [256k0], truncate oi.size to offset 256k, x3 = set_data_digest(crc32(x2)).
...
write is [4063232128k], write change to [40632320], truncate oi.size to offset 4063232, xn = set_data_digest(crs32(xn-1))
Now, we can see oi.size is 4063232, and data_digest is 0xffffffff, because thelength of in_data of crc is 0 every time.
when read verify crc will reply EIO. (EC pool).

so, when truncate in write, need clear data_digest and DIGEST flag,
when write before truncate, need to trim length, when offset over than oi.size, don't truncate oi.size to offset.

Fixes: https://tracker.ceph.com/issues/53240
Signed-off-by: jiawd [jiawendong@xtaotech.com](mailto:jiawendong@xtaotech.com)


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

ronen-fr commented on src/test/librados/testcase_cxx.cc:
```diff
@@ -389,10 +389,10 @@ void RadosTestECPP::SetUp()
-  ASSERT_TRUE(req);
-  ASSERT_EQ(0, ioctx.pool_required_alignment2(&alignment));
-  ASSERT_NE(0U, alignment);
+  //ASSERT_EQ(0, ioctx.pool_requires_alignment2(&req));
```
is the back and forth on the commenting out of these lines required?

NitzanMordhai replied:
@ronen-fr pushed that too soon, i fixed that.

ronen-fr commented:
@NitzanMordhai - commit test/pybind: check crc fail after append zero
isn't signed

NitzanMordhai commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -6706,13 +6706,14 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
                                              op.extent.truncate_size);
 	    }
+	    //truncate modify oi.size, need clear old data_digest and DIGEST flag
+	    oi.clear_data_digest();
```
@athanatos Jia, who wrote the original fix added that note: https://tracker.ceph.com/issues/53240#note-1
this is case 2, write before truncating

NitzanMordhai replied:
@athanatos agree, going to move it into the if

rzarzynski commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -6706,14 +6706,15 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
                                              oi,
                                              op.extent.truncate_size);
+	      //truncate modify oi.size, need clear old data_digest and DIGEST flag
+	      oi.clear_data_digest();
```
Oops, the commit title says " osd: Adjust code style" but this change isn't just a refactor – some extra in the git's stage area perhaps?

Anyway, I think they deserve a separate commit describing the intention.

NitzanMordhai replied:
done, move the line to the other commit

rzarzynski commented on src/test/librados/io_cxx.cc:
```diff
@@ -462,6 +462,19 @@ TEST_F(LibRadosIoPP, XattrListPP) {
 }
 
+TEST_F(LibRadosIoPP, CrcZeroWrite) {
+  char buf[128];
```
crimson lacks the fix, so I expect this test to fail there. Might we worth adding at a comment at least.

athanatos replied:
It's a simple fix -- can we just go ahead and fix it in crimson rather than letting the test fail?

NitzanMordhai replied:
actually, i tested it on crimson and it pass

petrutlucian94 replied:
This test fails sporadically with ENOTSUP:

```
/mnt/data/workspace/ceph_linux/src/test/librados/io_cxx.cc:888: Failure
Expected equality of these values:
  0
  ioctx.write("foo", bl, 0, sizeof(buf))
    Which is: -95
[  FAILED  ] LibRadosIoECPP.CrcZeroWrite (767 ms)
```

petrutlucian94 replied:
I'm thinking about skipping it on Windows for the time being: https://github.com/ceph/ceph-win32-tests/pull/19

rzarzynski commented on src/test/librados/io_cxx.cc:
```diff
@@ -462,6 +462,19 @@ TEST_F(LibRadosIoPP, XattrListPP) {
+  bufferlist bl;
+
+  ASSERT_EQ(0, ioctx.write("foo", bl, 0, 0));
+  ASSERT_EQ(0, ioctx.write("foo", bl, 0, sizeof(buf)));
```
Hmm, the only usage of the `0xcc`-filled `buf` seems to be the `sizeof()`. Is that truly the intention?

NitzanMordhai replied:
@rzarzynski yes, i had in mind to check that we didn't had any override above zero, but i didn't added it. i can remove the memset.

NitzanMordhai replied:
@rzarzynski please review

ljflores commented:
This PR had some unit test problems. Ref: https://trello.com/c/SFxA3U2X/1976-wip-yuri3-testing-2024-03-25-0744-old-wip-yuri3-testing-2024-03-24-1519-old-wip-yuri3-testing-2024-03-12-0851

NitzanMordhai commented:
> This PR had some unit test problems. Ref: https://trello.com/c/SFxA3U2X/1976-wip-yuri3-testing-2024-03-25-0744-old-wip-yuri3-testing-2024-03-24-1519-old-wip-yuri3-testing-2024-03-12-0851

the unit-test with EC needed allow_ec_overwrites, but since those test unit doesn't refresh the pool (delete\create) for each test, we can't remove the overwrites setting which causes the other unittest to fail.

rzarzynski commented:
jenkins test make check

rzarzynski commented:
jenkins retest this please

rzarzynski commented:
@NitzanMordhai: I'm seeing new changes. Should we repeat the QA?

NitzanMordhai commented:
> @NitzanMordhai: I'm seeing new changes. Should we repeat the QA?

i removed the new unit - test that I added for EC pools (keep only the non-EC pools tests), it won't work with truncate

ljflores commented:
@rzarzynski we will rerun tests.

rzarzynski commented:
jenkins test windows

rzarzynski commented:
jenkins test windows

rzarzynski commented:
QA in progress: https://tracker.ceph.com/issues/65349.

yuriw commented:
jenkins test windows

yuriw commented:
By @kamoltat "RADOS APPROVED https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues65349"

yuriw commented:
@NitzanMordhai pls merge when all checks passed

ref: https://tracker.ceph.com/issues/65349

NitzanMordhai commented:
jenkins test windows

NitzanMordhai commented:
jenkins test windows

NitzanMordhai commented:
jenkins retest this please

ljflores commented:
jenkins test api

ljflores commented:
jenkins test windows

NitzanMordhai commented:
jenkins test api

NitzanMordhai commented:
jenkins test api

petrutlucian94 commented:
I think the Windows failure was actually legit, one of the tests fails randomly with ``ENOTSUP``.

The ``make check`` job doesn't actually run this test.

rzarzynski commented:
@NitzanMordhai: we needs the commits as well in the reef backport.
